core/src/main/scala/kafka/server/AutoTopicCreationManager.scala

```diff
@@ -0,0 +1,139 @@
+
+  def apply(
+    config: KafkaConfig,
+    metadataCache: MetadataCache,
+    time: Time,
+    metrics: Metrics,
+    threadNamePrefix: Option[String],
+    adminManager: ZkAdminManager,
+    controller: KafkaController,
+    enableForwarding: Boolean
+  ): AutoTopicCreationManager = {
+
+    val channelManager =
+      if (enableForwarding)
+        Some(new BrokerToControllerChannelManager(
+          controllerNodeProvider = MetadataCacheControllerNodeProvider(
+            config, metadataCache),
+          time = time,
+          metrics = metrics,
+          config = config,
+          channelName = "autoTopicCreationChannel",
+          threadNamePrefix = threadNamePrefix,
+          retryTimeoutMs = config.requestTimeoutMs.longValue
+        ))
+      else
+        None
+    new AutoTopicCreationManagerImpl(channelManager, adminManager, controller, config.requestTimeoutMs)
+  }
+}
+
+class AutoTopicCreationManagerImpl(
+  channelManager: Option[BrokerToControllerChannelManager],
+  adminManager: ZkAdminManager,
+  controller: KafkaController,
+  requestTimeout: Int
+) extends AutoTopicCreationManager with Logging {
+
+  private val inflightTopics = new ConcurrentHashMap[String, CreatableTopic]
+
+  override def start(): Unit = {
+    channelManager.foreach(_.start())
+  }
+
+  override def shutdown(): Unit = {
+    channelManager.foreach(_.shutdown())
+  }
+
+  override def createTopics(topics: Set[CreatableTopic],
+                            controllerMutationQuota: ControllerMutationQuota): Unit = {
+    val topicConfigs = topics
+      .filter(topic => !inflightTopics.contains(topic.name()))
+      .map(topic => {(topic.name(), topic)}).toMap
+
+    if (topicConfigs.nonEmpty) {
+      if (!controller.isActive && channelManager.isDefined) {
+        // Mark the topics as inflight during auto creation through forwarding.
+        topicConfigs.foreach(config => inflightTopics.put(config._1, config._2))
+
+        val topicsToCreate = new CreateTopicsRequestData.CreatableTopicCollection
+        topicConfigs.foreach(config => topicsToCreate.add(config._2))
+        val createTopicsRequest = new CreateTopicsRequest.Builder(
+          new CreateTopicsRequestData()
+            .setTimeoutMs(requestTimeout)
+            .setTopics(topicsToCreate)
+        )
+
+        channelManager.get.sendRequest(createTopicsRequest, new ControllerRequestCompletionHandler {
+          override def onTimeout(): Unit = {
+            clearInflightRequests(topicConfigs)
+          }
+
+          override def onComplete(response: ClientResponse): Unit = {
+            clearInflightRequests(topicConfigs)
+          }
+        })
+      } else {
+        adminManager.createTopics(
+          requestTimeout,
+          validateOnly = false,
+          topicConfigs,
+          Map.empty,
+          controllerMutationQuota,
+          _ => ())
+      }
+    } else {
+      debug(s"Topics $topics are under creation already, skip sending additional " +
```

Comment:
We only get this message if _all_ of the topics are already inflight. Perhaps it is still useful if only some of them are inflight?

Reply:
Actually I don't think this logging is very useful, will replace it with something more explicit about state change.